### PR TITLE
Recipes for find_theorems and find_consts

### PR DESCRIPTION
**Pull request by lukasstevens**
Any feedback or additions? So far the recipe does not cover more than the reference manual.

### CODE REVIEW

File: src/commands/Find.thy [hunk 1]
```diff
@@ -0,0 +1,75 @@
+theory Find
+  imports Main
+begin
+
+section \<open>Find theorems with @{command find_theorems}\<close>
+text \<open>
+  Theorems can be searched by name where @{verbatim *} wildcards are allowed.
```
Comment (wimmers) [inline src/commands/Find.thy]: There is also the shorter form of this antiquotation, which makes the source more readable.
```suggestion
  Theorems can be searched by name where ▩‹*› wildcards are allowed.
```

File: src/commands/Find.thy [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  It is also possible to negate a pattern by prefixing it with a @{verbatim -}.
+  If a negated pattern occurs in a theorem, it will not be considered by the search.
+\<close>
+find_theorems "finite (?A::'a rel)"
+find_theorems "finite _ \<Longrightarrow> _" "insert _ _" -"card _"
+find_theorems "finite ?A \<Longrightarrow> finite ?B \<Longrightarrow> _ (?A \<union> ?B)"
+
+text \<open>
+  In addition to patterns, we can take the goal state into account to search for theorems.
+  The example below uses intro to search for introduction rules that solve the current goal.
```
Comment (wimmers) [inline src/commands/Find.thy]: I suggest we write ‹intro›, ‹elim›, ‹dest› and so on.
```suggestion
  The example below uses ‹intro› to search for introduction rules that solve the current goal.
```

File: src/commands/Find.thy [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  For the given goal, @{command find_theorems} finds the lemma @{thm bij_betw_byWitness} which we instantiate accordingly.
```
Comment (wimmers) [inline src/commands/Find.thy]: Typically, Isabelle users try to stay within the 100 characters line limit indicated by the blue line in jEdit.

File: src/commands/Find.thy [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  Then, we solve the the first two subgoals with @{method simp_all}.
+  This leaves us with two subgoals: @{prop \<open>(\<lambda>x. x + 1) ` \<int> \<subseteq> \<int>\<close>} and @{prop \<open>(\<lambda>x. x - 1) ` \<int> \<subseteq> \<int>\<close>}.
+  Searching with intro returns a lot of theorems since @{verbatim \<open>\<subseteq>\<close>} often occurs in the conclusion of theorems.
+  To filter out irrelevant theorems, we pass the goal as a pattern with the appropriate wildcards.
+  Then, @{command find_theorems} gives us the lemma @{thm image_subsetI} with which we can solve the remaining goals.
```
Comment (wimmers) [inline src/commands/Find.thy]:
```suggestion
  Then, @{command find_theorems} suggests the lemma @{thm image_subsetI} with which we can solve the remaining goals.
```

File: src/commands/Find.thy [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+section \<open>Find constants with @{command find_consts}\<close>
+text \<open>
+  Finding constants (e.g. functions or definitions) can be done with the @{command find_consts}.
+  As with @{command find_theorems}, you can search for constants by name as seen below.
+\<close>
+find_consts name: metric
+
+text \<open>
+  Perhaps more useful than searching constants by name, it is possible to search them by type.
```
Comment (wimmers) [inline src/commands/Find.thy]:
```suggestion
  Perhaps more useful than searching for constants by name, it is possible to search for them by type.
```

File: src/commands/Find.thy [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+begin
+
+section \<open>Find theorems with @{command find_theorems}\<close>
+text \<open>
+  Theorems can be searched by name where @{verbatim *} wildcards are allowed.
+  One of the theorems found with the query below is @{thm gauss_sum_nat}.
+\<close>
+find_theorems name: "gau*"
+
+text \<open>
```
Comment (wimmers) [inline src/commands/Find.thy:12]: Possible additions:
- searching by pattern and name can also be combined
- to look for infix operators like `<`, we either need to use the form `_ < _` or `(<)`

File: src/commands/Find.thy [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  Finding constants (e.g. functions or definitions) can be done with the @{command find_consts}.
+  As with @{command find_theorems}, you can search for constants by name as seen below.
+\<close>
+find_consts name: metric
+
+text \<open>
+  Perhaps more useful than searching constants by name, it is possible to search them by type.
+  The default search returns all constants which contain the given type pattern.
+  If you are only interested in exact matches, then you can use the strict annotation.
```
Comment (wimmers) [inline src/commands/Find.thy]: ‹strict›

**Comment by lukasstevens** [comment]
I applied all suggestions.

**Pull request merged**